compiler/syntax/src/jsx_v4.ml

```diff
@@ -1286,9 +1286,16 @@ let transform_structure_item ~config item =
       pstr_desc =
         Pstr_primitive ({pval_attributes; pval_type} as value_description);
     } as pstr -> (
-    match List.filter Jsx_common.has_attr pval_attributes with
-    | [] -> [item]
-    | [_] ->
+    match
+      ( List.filter Jsx_common.has_attr pval_attributes,
+        List.filter Jsx_common.has_attr_with_props pval_attributes )
+    with
+    | [], [] -> [item]
+    | _, [_] ->
+      Jsx_common.raise_error ~loc:pstr_loc
+        "Components cannot be defined as externals when using \
+         @react.componentWithProps. Please use @react.component instead."
```

Comment:
I think if the user intended to define a React component as a function of a props record, we should rather tell him to use the type `React.component<props>` in his external definition instead.

So maybe the message should be something like

```
Components cannot be defined as externals when using @react.componentWithProps.

If you intended to define an external for a React component using a props type,
use the type React.component<props> instead.
Alternatively, use @react.component for an external definition with labeled arguments.
```